server/sonar-server/src/main/java/org/sonar/server/platform/Platform.java

```diff
@@ -276,10 +305,63 @@ public Object getComponent(Object key) {
   }
 
   public enum Status {
-    BOOTING, SAFEMODE, UP
+    BOOTING, SAFEMODE, STARTING, UP
   }
 
   public enum Startup {
     NO_STARTUP_TASKS, ALL
   }
+
+  public interface PlatformAutoStarter {
+    /**
+     * Let the autostarted execute the provided code.
+     */
+    void execute(Runnable startCode);
+
+    /**
+     * This method is called by executed start code (see {@link #execute(Runnable)} has finished with a failure.
+     */
+    void failure(Throwable t);
+
+    /**
+     * This method is called by executed start code (see {@link #execute(Runnable)} has finished successfully.
+     */
+    void success();
+
+    /**
+     * Indicates whether the AutoStarter is running.
+     */
+    boolean isRunning();
+  }
+
+  private final class AsynchronousAutoStarter implements PlatformAutoStarter {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Make this a "static" inner class. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2694)